dubbo-common/src/main/java/org/apache/dubbo/common/URLBuilder.java

```diff
@@ -0,0 +1,342 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.common;
+
+import org.apache.dubbo.common.utils.CollectionUtils;
+import org.apache.dubbo.common.utils.StringUtils;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public final class URLBuilder {
+    private String protocol;
+
+    private String username;
+
+    private String password;
+
+    // by default, host to registry
+    private String host;
+
+    // by default, port to registry
+    private int port;
+
+    private String path;
+
+    private Map<String, String> parameters;
+
+    public URLBuilder() {
+        protocol = null;
+        username = null;
+        password = null;
+        host = null;
+        port = 0;
+        path = null;
+        parameters = new HashMap<>();
+    }
+
+    public URLBuilder(String protocol, String host, int port) {
+        this(protocol, null, null, host, port, null, null);
+    }
+
+    public URLBuilder(String protocol, String host, int port, String[] pairs) { // varargs ... conflict with the following path argument, use array instead.
+        this(protocol, null, null, host, port, null, CollectionUtils.toStringMap(pairs));
+    }
+
+    public URLBuilder(String protocol, String host, int port, Map<String, String> parameters) {
+        this(protocol, null, null, host, port, null, parameters);
+    }
+
+    public URLBuilder(String protocol, String host, int port, String path) {
+        this(protocol, null, null, host, port, path, null);
+    }
+
+    public URLBuilder(String protocol, String host, int port, String path, String... pairs) {
+        this(protocol, null, null, host, port, path, CollectionUtils.toStringMap(pairs));
+    }
+
+    public URLBuilder(String protocol, String host, int port, String path, Map<String, String> parameters) {
+        this(protocol, null, null, host, port, path, parameters);
+    }
+
+    public URLBuilder(String protocol, String username, String password, String host, int port, String path, Map<String, String> parameters) {
+        this.protocol = protocol;
+        this.username = username;
+        this.password = password;
+        this.host = host;
+        this.port = port;
+        this.path = path;
+        this.parameters = parameters != null ? parameters : new HashMap<>();
+    }
+
+    public URLBuilder(URL url) {
+        protocol = url.getProtocol();
+        username = url.getUsername();
+        password = url.getPassword();
+        host = url.getHost();
+        port = url.getPort();
+        path = url.getPath();
+        parameters = new HashMap<>(url.getParameters());
+    }

```

Comment:
I think providing a method like`URLBuilder.from(url)` would be better than using a constructor directly.

What do you think?

Reply:
updated as requested